gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +436,81 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
     return null;
   }
 
+  @VisibleForTesting
+  ChannelCredentials createMtlsToS2AChannelCredentials(
+      InputStream trustBundle, InputStream privateKey, InputStream certChain) throws IOException {
+    if (trustBundle == null || privateKey == null || certChain == null) {
+      return null;
+    }
+    return TlsChannelCredentials.newBuilder()
+        .keyManager(privateKey, certChain)
+        .trustManager(trustBundle)
+        .build();
+  }
+
+  ChannelCredentials createPlaintextToS2AChannelCredentials(String plaintextAddress) {
+    if (Strings.isNullOrEmpty(plaintextAddress)) {
+      return null;
+    }
+    return S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
+        .build();
+  }
+
+  /**
+   * This method creates gRPC {@link ChannelCredentials} configured to use S2A to estbalish a mTLS
+   * connection. First, the address of S2A is discovered by using the {@link S2A} utility to learn
+   * the {@code mtlsAddress} to reach S2A and the {@code plaintextAddress} to reach S2A. Prefer to
+   * use the {@code mtlsAddress} address to reach S2A if it is non-empty and the MTLS-MDS
+   * credentials can successfully be discovered and used to create {@link TlsChannelCredentials}. If
+   * there is any failure using mTLS-to-S2A, fallback to using a plaintext connection to S2A using
+   * the {@code plaintextAddress}.
```

Comment:
nit: `createPlaintextToS2AChannelCredentials` can return null which will also be treated differently.

Can you add that if there is no plaintextAddress supplied (null or empty), then plaintext connection is not possible. If you know any of the error scenarios for not having a plaintext address, that would be great too!

The last option is this one: `ManagedChannelBuilder.forAddress(serviceAddress, port);` (I'm not sure if there is a term for this)

Reply:
Re: The last option is this one: ManagedChannelBuilder.forAddress(serviceAddress, port); (I'm not sure if there is a term for this)
> This option is what we go to if `endpointContext.useS2A() == false` or if mtls-to-s2a fails and plaintext address is empty. This is a TLS connection to the `serviceAddress`. Looking at how `serviceAddress` is computed, I think we should this same thing from `mtlsEndpoint`.

In https://github.com/googleapis/sdk-platform-java/pull/3326/commits/a7af12e44b987e2edd7a7f7a1113a14cf5527001 updated javadoc comment, and added `mtlsServiceAddress` (created from `mtlsEndpoint`)

Reply:
Actually looking at this again, we should be using mtlsEndpoint not mtlsServiceAddress, reverted in commit: